File: samples/grpc/iot-service-scala/src/main/scala/iot/registration/RegistrationEvents.scala
package iot.registration

import scala.concurrent.Future

import akka.actor.typed.ActorSystem
import akka.http.scaladsl.model.HttpRequest
import akka.http.scaladsl.model.HttpResponse
import akka.persistence.query.typed.EventEnvelope
import akka.projection.grpc.producer.EventProducerSettings
import akka.projection.grpc.producer.scaladsl.EventProducer
import akka.projection.grpc.producer.scaladsl.EventProducer.Transformation

object RegistrationEvents {

  def eventProducerService(system: ActorSystem[_])
      : PartialFunction[HttpRequest, Future[HttpResponse]] = {
    val transformation = Transformation.empty
      .registerAsyncEnvelopeMapper[Registration.Registered, proto.Registered](
        envelope => Future.successful(Some(transformRegistered(envelope))))

    val eventProducerSource = EventProducer.EventProducerSource(
      Registration.EntityKey.name,
      "registration-events",
      transformation,
      EventProducerSettings(system))

    EventProducer.grpcServiceHandler(eventProducerSource)(system)
  }

  private def transformRegistered(
      envelope: EventEnvelope[Registration.Registered]): proto.Registered = {
    val event = envelope.event
    proto.Registered(secret = Some(proto.SecretDataValue(event.secret.value)))
  }

}

File: samples/grpc/iot-service-java/src/main/java/iot/registration/RegistrationEvents.java
package iot.registration;

import akka.actor.typed.ActorSystem;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.japi.function.Function;
import akka.projection.grpc.producer.EventProducerSettings;
import akka.projection.grpc.producer.javadsl.EventProducer;
import akka.projection.grpc.producer.javadsl.EventProducerSource;
import akka.projection.grpc.producer.javadsl.Transformation;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

public class RegistrationEvents {

  public static Function<HttpRequest, CompletionStage<HttpResponse>> eventProducerService(
      ActorSystem<?> system) {
    Transformation transformation =
        Transformation.empty()
            .registerMapper(Registration.Registered.class, RegistrationEvents::transformRegistered);

    EventProducerSource eventProducerSource =
        new EventProducerSource(
            Registration.ENTITY_KEY.name(),
            "registration-events",
            transformation,
            EventProducerSettings.create(system));

    return EventProducer.grpcServiceHandler(system, eventProducerSource);
  }

  private static Optional<iot.registration.proto.Registered> transformRegistered(
      Registration.Registered event) {
    return Optional.of(
        iot.registration.proto.Registered.newBuilder()
            .setSecret(
                iot.registration.proto.SecretDataValue.newBuilder()
                    .setValue(event.secret.value)
                    .build())
            .build());
  }
}
